src/Microsoft.AspNetCore.SystemWebAdapters.FrameworkServices/ProxyHeaderModule.cs

```diff
@@ -103,6 +110,31 @@ private static void UseForwardedFor(NameValueCollection requestHeaders, NameValu
         }
     }
 
+    private class ReflectionHelper
+    {
+        private readonly FieldInfo _wrField;
+        private readonly Type _iis7WorkerRequestType;
+        private readonly FieldInfo _knownRequestHeadersField;
+
+        public ReflectionHelper()
+        {
+            _wrField = typeof(HttpRequest).GetField("_wr", BindingFlags.NonPublic | BindingFlags.Instance);
+            _iis7WorkerRequestType = Assembly.GetAssembly(typeof(HttpRequest)).GetType("System.Web.Hosting.IIS7WorkerRequest");
+            _knownRequestHeadersField = _iis7WorkerRequestType.GetField("_knownRequestHeaders", BindingFlags.NonPublic | BindingFlags.Instance);
+        }
+        public void SetHostHeader(HttpRequest? request, string host)
+        {
+            // Need to use reflection to force HttpRequest to update the known headers
+            if (request != null)
+            {
+                var workerRequest = _wrField.GetValue(request);
+                var knownRequestHeaders = (string[])_knownRequestHeadersField.GetValue(workerRequest);
```

Comment:
is there any chance this will be null?

Reply:
In theory, yes. It's `null` until the headers are accessed the first time. This normally happens *very* early in the pipeline, during request validation, but I think there might be a way to turn off request validation, even if it's super frowned upon 😅

Reply:
sounds like we should probably check for null before updating it. If it's null, then it should propagate fine and this workaround wouldn't be needed, right?

Reply:
Yeah, if it's null, it can just return early as the correct value will be read the first time 👍

Reply:
Yeah if its null we don't need to run this at all, so good idea to just do nothing if we don't get the array